velox/common/memory/tests/SharedArbitratorTest.cpp

```diff
@@ -426,6 +431,148 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, reclaimFromOrderBy) {
+  std::vector<RowVectorPtr> vectors;
+  for (int i = 0; i < numVectors; ++i) {
+    vectors.push_back(newVector());
+  }
+  createDuckDbTable(vectors);
+
+  std::atomic_bool task1TestValueHookHit{false};
+  folly::EventCount task2CreationWait;
+  std::atomic_bool task1CreateDriverResume{false};
+  folly::EventCount task1CreateDriverWait;
+  std::atomic_bool task2AllocCbResume{false};
+  folly::EventCount task2AllocCbWait;
+  std::shared_ptr<core::QueryCtx> queryCtx = newQueryCtx(
+      2 * MB, std::make_unique<TestMemoryReclaimer>([&](MemoryPool* pool) {
+        if (pool->reservedBytes() == 0) {
+          task2AllocCbResume = true;
+          task2AllocCbWait.notifyAll();
+        }
+      }));
+
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::Task::createDriversLocked",
+      std::function<void(void*)>([&](void* /*unused*/) {
+        task1TestValueHookHit = true;
+        task2CreationWait.notifyAll();
+        task1CreateDriverWait.await(
+            [&]() { return task1CreateDriverResume.load(); });
+      }));
+  std::thread task1Thread([&]() {
+    auto task1 =
+        AssertQueryBuilder(duckDbQueryRunner_)
+            .queryCtx(queryCtx)
+            .plan(
+                PlanBuilder()
+                    .values(vectors)
+                    .project(
+                        {"c0 / c1 as x",
+                         "c1 / c0 as y",
+                         "'need 12 chars to be larger than inline size (StringView::kInlineSize) to trigger pool allocation from expr compiling' as z"})
+                    .planNode())
+            .assertResults(
+                "SELECT c0 / c1 as x, c1 / c0 as y, 'need 12 chars to be larger than inline size (StringView::kInlineSize) to trigger pool allocation from expr compiling' as z FROM tmp");
+  });
+
+  fakeOperatorFactory_->setAllocationCallback([&](Operator* op) {
+    auto buffer = op->pool()->allocate(1 * MB);
+    task1CreateDriverResume = true;
+    task1CreateDriverWait.notifyAll();
+    task2AllocCbWait.await([&]() { return task2AllocCbResume.load(); });
+    return TestAllocation{op->pool(), buffer, 1 * MB};
+  });
+  std::thread task2Thread([&]() {
+    task2CreationWait.await([&]() { return task1TestValueHookHit.load(); });
+    TestValue::clear("facebook::velox::exec::Task::createDriversLocked");
+    try {
+      auto task2 =
+          AssertQueryBuilder(duckDbQueryRunner_)
+              .queryCtx(queryCtx)
+              .plan(PlanBuilder()
+                        .values(vectors)
+                        .addNode([&](std::string id, core::PlanNodePtr input) {
+                          return std::make_shared<FakeMemoryNode>(id, input);
+                        })
+                        .planNode())
+              .assertResults("SELECT * FROM tmp");
+    } catch (const VeloxRuntimeError& e) {
+      VELOX_CHECK_EQ(e.errorCode(), error_code::kMemCapExceeded);
+    }
+  });
+  waitForAllTasksToBeDeleted(10'000'000);
+  task1Thread.join();
```

Comment:
Let's first wait test thread to joins and then wait for tasks to complete? Thanks!

Reply:
Then in a fault situation it will hang.